dotnet/src/Connectors/Connectors.Google/GeminiToolCallBehavior.cs

```diff
@@ -0,0 +1,189 @@
+
+    /// <summary>Initializes the instance; prevents external instantiation.</summary>
+    private GeminiToolCallBehavior(bool autoInvoke)
+    {
+        this.MaximumAutoInvokeAttempts = autoInvoke ? DefaultMaximumAutoInvokeAttempts : 0;
+    }
+
+    /// <summary>Gets how many requests are part of a single interaction should include this tool in the request.</summary>
+    /// <remarks>
+    /// This should be greater than or equal to <see cref="MaximumAutoInvokeAttempts"/>. It defaults to <see cref="int.MaxValue"/>.
+    /// Once this limit is reached, the tools will no longer be included in subsequent retries as part of the operation, e.g.
+    /// if this is 1, the first request will include the tools, but the subsequent response sending back the tool's result
+    /// will not include the tools for further use.
+    /// </remarks>
+    public int MaximumUseAttempts { get; } = int.MaxValue;
+
+    /// <summary>Gets how many tool call request/response roundtrips are supported with auto-invocation.</summary>
+    /// <remarks>
+    /// To disable auto invocation, this can be set to 0.
+    /// </remarks>
+    public int MaximumAutoInvokeAttempts { get; }
+
+    /// <summary>
+    /// Gets whether validation against a specified list is required before allowing the model to request a function from the kernel.
+    /// </summary>
+    /// <value>true if it's ok to invoke any kernel function requested by the model if it's found;
+    /// false if a request needs to be validated against an allow list.</value>
+    internal virtual bool AllowAnyRequestedKernelFunction => false;
+
+    /// <summary>Configures the <paramref name="request"/> with any tools this <see cref="GeminiToolCallBehavior"/> provides.</summary>
+    /// <param name="kernel">The <see cref="Kernel"/> used for the operation.
+    /// This can be queried to determine what tools to provide into the <paramref name="request"/>.</param>
+    /// <param name="request">The destination <see cref="GeminiRequest"/> to configure.</param>
+    internal abstract void ConfigureGeminiRequest(Kernel? kernel, GeminiRequest request);
+
+    internal GeminiToolCallBehavior Clone()
+    {
+        return (GeminiToolCallBehavior)this.MemberwiseClone();
+    }
+
+    /// <summary>
+    /// Represents a <see cref="GeminiToolCallBehavior"/> that will provide to the model all available functions from a
+    /// <see cref="Kernel"/> provided by the client.
+    /// </summary>
+    internal sealed class KernelFunctions : GeminiToolCallBehavior
+    {
+        internal KernelFunctions(bool autoInvoke) : base(autoInvoke) { }
+
+        public override string ToString() => $"{nameof(KernelFunctions)}(autoInvoke:{this.MaximumAutoInvokeAttempts != 0})";
+
+        internal override void ConfigureGeminiRequest(Kernel? kernel, GeminiRequest request)
+        {
+            // If no kernel is provided, we don't have any tools to provide.
+            if (kernel is null)
+            {
+                return;
```

Comment:
Below in `EnabledFunctions` if there is no kernel and auto invoke is set to true an exception is thrown. Why is the behaviour different in this case?

Reply:
OpenAI impl also throws exception here https://github.com/microsoft/semantic-kernel/blob/79620c727e61d056736674018156715cba690a92/dotnet/src/Connectors/Connectors.OpenAI/ToolCallBehavior.cs#L192

Reply:
@Krzysztof318, that was a good spot from @markwallace-microsoft, we should have consistency on those. 
By looking into this #5655, the pattern is to thrown an exception if no `kernel` is provided.

Reply:
@RogerBarreto @markwallace-microsoft  this PR #5637 only changes RequiredFunctions (gemini doesn't have this). Only has KernelFunction and EnabledFunction.
EnabledFunction throws exception if kernel is null, this is consistent with openAI ToolCallBehaviour.
KernelFunction doesn't throw exception if kernel is null, this is also consistent with openAI ToolCallBehaviour.